pennylane/qnodes/base.py

```diff
@@ -264,6 +266,55 @@ def qfunc(a, w):
             "The QNode can only be drawn after its CircuitGraph has been constructed."
         )
 
+    def set_trainable_args(self, arg_indices):
+        """Store the indices of quantum function positional arguments
+        that support differentiability.
+
+        Args:
+            args (None or Set[int]): Differentiable positional argument indices. A
+                value of ``None`` means that all argument indices are differentiable.
+        """
+        if not self.mutable and self.circuit is not None:
+
+            if self.get_trainable_args() == arg_indices:
+                return
+
+            raise QuantumFunctionError(
+                "The trainability of arguments on immutable QNodes cannot be modified after the first evaluation."
+            )
+
+        if not arg_indices:
+            # the provided arg_indices are an empty set
+            self._trainable_args = None
+            return
+
+        # Perform validation
+        if not self.func.var_pos and max(arg_indices) > self.func.n_pos:
+            # QNode does not allow variable positional arguments (*args), and
+            # the provided index set contains a value larger than the number
+            # of positional arguments.
```

Comment:
A bit confused here :thinking: 
Wouldn't passing  a value larger than the number of variable positional args be ok, since it is only the positional args that are not allowed to be variables in this case?

Reply:
I don't think I worded it very well 🤔 

This validation is to catch the case where we might have a QNode defined as `circuit(x, y, z)`, but we try to do `circuit.set_trainable_args = {0, 1, 6}`.

However, if we have a QNode defined as `circuit(*args)`, then:

* `self.func.var_pos == True` since `*args` is present, but
* `self.func.n_pos == 1` because function introspection looks at the signature and counts `*args` as a single argument.

So if a QNode is defined using `*args`, it becomes impossible to perform this validation, since we don't know how many arguments the function is supposed to have. This follows even in the general case `circuit(x, y, z, *args)`.